numpy/core/code_generators/numpy_api.py

```diff
@@ -350,6 +350,9 @@
     'PyArray_ResolveWritebackIfCopy':       (302,),
     'PyArray_SetWritebackIfCopyBase':       (303,),
     # End 1.14 API
+    'NpyIter_SafeDealloc':                  (304,),
+    'NpyIter_ErrorOccurred':                (305,),
```

Comment:
I understand we can do without `NpyIter_SafeDealloc`. Do we need to export `NpyIter_ErrorOccurred` or can it remain an internal API?

Reply:
I am not sure what to do best. We can definitely do without the first, yes.  For the second one we need to figure out what to do with the error which is set on the fly (possibly when the API is not held).

The current solution is to stash the error and raise it later when we (probably) have the API and can return error values, because `iternext` does not report error returns.  I think we will need to stash the error quite often to cleanup correctly, but that does not have to be done by NpyIter itself in principle.  It may be possible to just flag an error return, I think this was my first reaction, because it can't be disrupted e.g. by a random `PyErr_Clear()` call.

The issue is that the `iternext` functions use `0` to indicate stop-iteration so we can't distinguish an aborted iteration (due to an error) from a finished iteration.  That is why I needed to add `NpyIter_ErrorOccurred()`, there needs to be a way to know that the iteration was aborted when `dealloc` is not called immediately.  That is mostly important for `np.nditer` (the python API).

As to the `NpyIter_SafeDealloc`, no I am not sure its the best design. If it seems nicer than the alternative, we could use it privately I suppose.  `NpyIter_Deallocate` may need the Python API, so as I wrote it here, it has to either check `PyErr_Occurred()` or the caller has to stash the error.

Right now of course, we ignore that this type of thing can happen. In practice `NpyIter_Dealloc` is simply called with an error already set, and we basically hope it will all work out and the correct error is raised (possibly much belated).

Reply:
maybe @eric-wieser has some thoughts

Reply:
Hmmm, there is probably a remaining hole here. When the iteration is never finished, either due to an error or because `np.nditer` is just not iterated to the end, we need to clean up buffers in that case as well. Technically, we should copy the buffer back if (but I guess only if) there was no error. Also an existing issue.

I guess right now we simply, assume that `NpyIter_Deallocate` can handle such things (its called by `np.nditer.__exit__`). Probably the `Reset` functions are in a similar spot.